VERSIONING.md

```diff
@@ -0,0 +1,215 @@
+      be versioned with a major version greater than `v0`.
+    * Experimental modules will start their versioning at `v0.0.0` and will
+      increment their minor version when backwards incompatible changes are
+      released and increment their patch version when backwards compatible
+      changes are released.
+    * Stable contrib modules cannot depend on experimental modules from this
+      project.
+    * All stable contrib modules of the same major version with this project
+      will use the same entire version as this project.
+      * Stable modules may be released with an incremented minor or patch
+        version even though that module's code has not been changed. Instead
+        the only change that will have been included is to have updated that
+        modules dependency on this project's stable APIs.
+      * When an experimental module in contrib becomes stable a new stable
+        module version will be released and will include this now stable
+        module. The new stable module version will be an increment of the minor
+        version number and will be applied to all existing stable contrib
+        modules, this project's modules, and the newly stable module being
+        released.
+  * Contrib modules will be kept up to date with this project's releases.
+    * Due to the dependency contrib modules will implicitly have on this
+      project's modules the release of stable contrib modules to match the
+      released version number will be staggered after this project's release.
+      There is no explicit time guarantee for how long after this projects
+      release the contrib release will be. Effort should be made to keep them
+      as close in time as possible.
+    * No additional stable release in this project can be made until the
+      contrib repository has a matching stable release.
+    * No release can be made in the contrib repository after this project's
+      stable release except for a stable release of the contrib repository.
+* GitHub releases will be made for all releases.
+* Go modules will be made available at Go package mirrors.
+
+## Example Versioning Lifecycle
+
+To better understand the implementation of the above policy the following
+example is provided. This project is simplified to include only the following
+modules and their versions:
+
+* `otel`: `v0.14.0`
+* `otel/trace`: `v0.14.0`
+* `otel/metrics`: `v0.14.0`
+* `otel/baggage`: `v0.14.0`
+* `otel/sdk/trace`: `v0.14.0`
+* `otel/sdk/metrics`: `v0.14.0`
+
+These modules have been developed to a point where the `otel/trace`,
+`otel/baggage`, and `otel/sdk/trace` modules have reached a point that they
+should be considered for a stable release. The `otel/metrics` and
+`otel/sdk/metrics` are still under active development and the `otel` module
+depends on both `otel/trace` and `otel/metrics`.
+
+The `otel` package is refactored to remove its dependencies on `otel/metrics`
```

Comment:
This point leaves me with a question if this document should also make a policy on how do we decide if a module should become stable or not. If we assume the situation in this example so far (trace stable, metric experimental), should toplevel module become stable (thus moving/removing a functionality depending on metric) or should stay experimental as long as metric is?

The former policy (make stable, drop experimental code) has a risk of adding awkward code if the module has an interface that uses types or even mentions things from both stable and unstable. I haven't come up with a scenario where awkward code might even be breaking compatibility, but maybe this is also something to be considered.

The latter policy (keep experimental as long as any dependency is experimental) might lead to toplevel module or exporter modules to wait a really long time before reaching stability (like we don't even have logs functionality).

Reply:
> The former policy (make stable, drop experimental code) has a risk of adding awkward code if the module has an interface that uses types or even mentions things from both stable and unstable. I haven't come up with a scenario where awkward code might even be breaking compatibility, but maybe this is also something to be considered.

That is a good point, adding a method to an interface is a breaking change. If there is a conglomerate interface that includes things from an experimental package, when it is broken apart to allow the module to become stable it cannot be added back in as a part of that interface.

I don't think we have instances of this currently, but it is going to be important to keep this in mind going forward.

> The latter policy (keep experimental as long as any dependency is experimental) might lead to toplevel module or exporter modules to wait a really long time before reaching stability (like we don't even have logs functionality).

This approach would not allow compatibility with new development. If we did wait for metrics to become stable before making `otel` stable it would not be able to account for new signals like logs, events, or errors. That would mean that the `otel` package would never go stable or have to fluctuate back and forth.

Reply:
> > The former policy (make stable, drop experimental code) has a risk of adding awkward code if the module has an interface that uses types or even mentions things from both stable and unstable. I haven't come up with a scenario where awkward code might even be breaking compatibility, but maybe this is also something to be considered.
> 
> That is a good point, adding a method to an interface is a breaking change. If there is a conglomerate interface that includes things from an experimental package, when it is broken apart to allow the module to become stable it cannot be added back in as a part of that interface.
> 
> I don't think we have instances of this currently, but it is going to be important to keep this in mind going forward.

We kinda-sorta may have it with `ProtocolDriver` interface in `exporters/otlp` (#1369). I suppose that in this case, we will need to create an extra interface, just like in golang's `database/sql/driver` there is a main interface `Conn` and some extra interfaces that a `Conn` implementation may also implement (like `Execer`, `ExecerContext` or `ConnBeginTx`).

Still, should the document have the wording about a process of declaring a stable module? Even if it's one sentence about it being a case-by-case thing.

Reply:
> We kinda-sorta may have it with `ProtocolDriver` interface in `exporters/otlp` (#1369). I suppose that in this case, we will need to create an extra interface, just like in golang's `database/sql/driver` there is a main interface `Conn` and some extra interfaces that a `Conn` implementation may also implement (like `Execer`, `ExecerContext` or `ConnBeginTx`).

Yeah, I think you described the development process I had envisioned.

> Still, should the document have the wording about a process of declaring a stable module? Even if it's one sentence about it being a case-by-case thing.

[Added a sentence](https://github.com/open-telemetry/opentelemetry-go/pull/1388/commits/3d481e69fd58aa0786bc2d2a5914c41992a50ab9):

> The decision to make a module stable will be made on a case-by-case bases by the maintainers of this project.